core/electrum_client/src/types.rs

```diff
@@ -0,0 +1,201 @@
+            result.swap(i, 31 - i);
+        }
+
+        result
+    }
+}
+
+fn from_hex<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where
+    T: FromHex,
+    D: de::Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    T::from_hex(&s).map_err(de::Error::custom)
+}
+
+fn from_hex_array<'de, T, D>(deserializer: D) -> Result<Vec<T>, D::Error>
+where
+    T: FromHex + std::fmt::Debug,
+    D: de::Deserializer<'de>,
+{
+    let arr = Vec::<String>::deserialize(deserializer)?;
+
+    let results: Vec<Result<T, _>> = arr
+        .into_iter()
+        .map(|s| T::from_hex(&s).map_err(de::Error::custom))
+        .collect();
+
+    let mut answer = Vec::new();
+    for x in results.into_iter() {
+        answer.push(x?);
+    }
+
+    Ok(answer)
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetHistoryRes {
+    pub height: i32,
+    pub tx_hash: Txid,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ListUnspentRes {
+    pub height: usize,
+    pub tx_pos: usize,
+    pub value: u64,
+    pub tx_hash: Txid,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ServerFeaturesRes {
+    pub server_version: String,
+    #[serde(deserialize_with = "from_hex")]
+    pub genesis_hash: [u8; 32],
+    pub protocol_min: String,
+    pub protocol_max: String,
+    pub hash_function: Option<String>,
+    pub pruning: Option<i64>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetHeadersRes {
+    pub max: usize,
+    pub count: usize,
+    #[serde(rename(deserialize = "hex"), deserialize_with = "from_hex")]
+    pub raw_headers: Vec<u8>,
+    #[serde(skip)]
+    pub headers: Vec<block::BlockHeader>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetBalanceRes {
+    pub confirmed: u64,
+    pub unconfirmed: u64,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetMempoolRes {
+    pub fee: u64,
+    pub height: i32,
+    pub tx_hash: Txid,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetMerkleRes {
+    pub block_height: usize,
+    pub pos: usize,
+    #[serde(deserialize_with = "from_hex_array")]
+    pub merkle: Vec<[u8; 32]>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct HeaderNotification {
+    pub height: usize,
+    #[serde(rename(serialize = "hex"))]
+    pub header: block::BlockHeader,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ScriptNotification {
+    pub scripthash: ScriptHash,
+    pub status: ScriptStatus,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    IOError(std::io::Error),
+    JSON(serde_json::error::Error),
+    Hex(bitcoin::hashes::hex::Error),
+    Protocol(serde_json::Value),
+    Bitcoin(bitcoin::consensus::encode::Error),
+    AlreadySubscribed(ScriptHash),
+    NotSubscribed(ScriptHash),
+    InvalidResponse(serde_json::Value),
+    Message(String),
+
+    #[cfg(feature = "ssl")]
+    InvalidSslMethod(openssl::error::ErrorStack),
+    #[cfg(feature = "ssl")]
+    SslHandshakeError(openssl::ssl::HandshakeError<std::net::TcpStream>),
+}
+
+impl std::convert::From<std::io::Error> for Error {
```

Comment:
nit: I would use a macro. Probably overkill, but could be useful in future if we'll have more cases to handle.